text/0000-router-service-computed-query-params.md

```diff
@@ -0,0 +1,324 @@
+
+
+### Sticky Query Params
+
+By default, transitionTo will clear the query params, unless specified inside the transiion.
+
+If query params are defined ahead of time as sticky, they will persist in the URL between sub routes.
+
+This can be configured in the `Router.map` function:
+```ts
+Router.map(function() {
+  this.queryParams('bar');
+
+  this.route('faq');
+
+  this.route('posts', function() {
+    this.queryParams('foo', 'baz');
+
+    this.route('new');
+    this.route('index');
+    this.route(
+      'show',
+      { path: '/:postId', queryParams: ['hideComments', 'invertColors'] },
+      function() {
+        this.route('edit');
+        this.route('comment');
+        this.route('share');
+      }
+    );
+  });
+});
+```
+
+This method of dealing with query params implies the following:
+ - All query params, even if not specified are allowed. given the above example, I could use the qp "strongestAvenger" anywhere
+ - Any sort of transition will clear the query params, unless it is defined as a sticky queryParam. So, if I'm on the posts/new route with the query params "foo" and "baz", and transition to posts/show, those query params are still available. If I navigate to the faq page, foo and baz are cleared from the URL.
+ - The globally defined query params will stick around until cleared manually. If I visit faq?bar=2, and then transition to posts. the bar=2 query param will still be present.
+ - The `this.queryParams` function will be available at every nesting of the route tree, but `queryParams` are also configurable in the route options hash.
+
+Notes:
+ - This does not imply a caching mechanism (like what controllers today allow)
+ - If a certain app wants caching of query params as they exist today, an addon could be made to fill that gap.
+
+ **A more concrete example: filter query params**
+Given we want a way to search over a list of products, be able to view additional information about a selected search result, and save the search result for later, sticky params help maintain the search throughout all of the route navigations (similar to Amazon). 
+
+
+```ts
+Router.map(function() {
+  this.queryParams('bar');
+
+  this.route('search', { queryParams: [
+    'term', 'isPrime', 'department', 'averageReview', 'brand', 
+    'memoryType', 'processor', 'vRamCapacity', 'certification',
+  ]}, function() {
+    // index route will be the search results page
+
+    // shows a selected result with additional information
+    this.route('summary', { path: '/summary/:itemId' });
+
+    // shows a modal with a field to name the search to be loaded later
+    this.route('save');
+  });
+});
+```
+
+1. visit `/search`.
+2. type "RTX" into the search bar press the enter key.
+3. the URL now shows `/search?term=RTX` and some results display as rows.
+   The code for appending `term` may look like the following:
+
+   ```ts
+   @service router;
+
+   this.router.setQueryParam('term', 'RTX');
```

Comment:
An alternative here is to have `this.router.queryParams.set('term', 'RTX')`. I get that there's value in `queryParams` being a POJO, and that introducing a custom type here may introduce hurdles for people implementing their own `serialize`/`deserialize` hooks, but defining a more strict interface may also be a good thing in the long run (similar to `URLSearchParams`, for example)

Reply:
I'd like to stay away from set/get as they're not really present in idiomatic Octane.

> but defining a more strict interface may also be a good thing in the long run

I think being simple first is more important here, as the current query params are too restrictive and there are too many unknowns with (de)serialization